### PR TITLE
[Doc] Update FAQ doc about binary segmentation and ReduceZeroLabel

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation

Update FAQ doc about  binary segmentation problem and `reduce_zero_label` usage.

### CODE REVIEW

File: docs/en/faq.md [hunk 1]
```diff
@@ -66,3 +66,105 @@ In the test script, we provide `show-dir` argument to control whether output the
...
+    seg_pred = seg_logit.argmax(dim=1)
+```
+
+- When `out_channels=2`, using Cross Entropy Loss in training, using `F.softmax()` and `argmax()` to get prediction of each pixel in inference.
+
+- When `out_channels=1`, we provide a parameter `threshold(default to 0.3)` in [#2016](https://github.com/open-mmlab/mmsegmentation/pull/2016), using Binary Cross Entropy Loss in training, using `F.sigmoid()` and `threshold` to get prediction of each pixel in inference.
+
+More details about calculating segmentation prediction could be found in [encoder_decoder.py](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py):
+
+In summary, to implement binary segmentation methods users should modify below parameters in config files:
```
Comment (xiexinch) [inline docs/en/faq.md:130]:
```suggestion
In summary, to implement binary segmentation methods users should modify the below parameters in the `decode_head` and `auxiliary_head` configs:
```

File: docs/zh_cn/faq.md [hunk 2]
```diff
@@ -66,3 +66,104 @@
...
+  self.threshold = threshold
+  ...
+  self.conv_seg = nn.Conv2d(channels, self.out_channels, kernel_size=1)
+```
+
+有两种计算二值分割任务的方法:
+
+- 当 `out_channels=2` 时, 在训练时以 Cross Entropy Loss 作为损失函数, 在推理时使用 `F.softmax()` 归一化 logits 值, 然后通过 `argmax()` 得到每个像素的预测结果.
+
+- 当 `out_channels=1` 时, 我们在 [#2016](https://github.com/open-mmlab/mmsegmentation/pull/2016) 里提供了阈值参数 `threshold (默认为 0.3)`, 在训练时以 Binary Cross Entropy Loss 作为损失函数, 在推理时使用 `F.sigmoid()` 和在 `threshold` 得到预测结果.
```
Comment (xiexinch) [inline docs/zh_cn/faq.md:110]:
```suggestion
- 当 `out_channels=1` 时, 我们在 [#2016](https://github.com/open-mmlab/mmsegmentation/pull/2016) 里提供了阈值参数 `threshold (默认为 0.3)`, 在训练时以 Binary Cross Entropy Loss 作为损失函数, 在推理时使用 `F.sigmoid()` 和 `threshold` 得到预测结果.
```

File: docs/zh_cn/faq.md [hunk 3]
```diff
@@ -66,3 +66,104 @@
...
+if self.out_channels == 1:
+    seg_pred = (seg_logit >
+                self.decode_head.threshold).to(seg_logit).squeeze(1)
+else:
+    seg_pred = seg_logit.argmax(dim=1)
+```
+
+更多关于计算语义分割预测的细节可以参考 [encoder_decoder.py](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py):
+
+对于实现上述两种计算二值分割的方法, 需要分别在配置文件里修改:
```
Comment (xiexinch) [inline docs/zh_cn/faq.md:130]:
```suggestion
对于实现上述两种计算二值分割的方法, 需要在 `decode_head` 和 `auxiliary_head` 的配置里修改:
```

**Review by xiexinch — APPROVED** [review]

File: docs/zh_cn/faq.md [hunk 4]
```diff
@@ -66,3 +66,104 @@
...
+  if out_channels == 1 and threshold is None:
+      threshold = 0.3
+      warnings.warn('threshold is not defined for binary, and defaults'
+                    'to 0.3')
+  self.num_classes = num_classes
+  self.out_channels = out_channels
+  self.threshold = threshold
+  ...
+  self.conv_seg = nn.Conv2d(channels, self.out_channels, kernel_size=1)
+```
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:104]: please remove these lines and add 
```suggestion
`num_classes` 应该和数据集本身类别个数一致，当是二值分割时，数据集只有前景和背景两类，所以 `num_classes` 为 2。`out_channels` 控制模型最后一层的输出的通道数，通常和 `num_classes` 相等，但当二值分割时候，可以有两种处理方法，分别是：
```

File: docs/zh_cn/faq.md [hunk 5]
```diff
@@ -66,3 +66,104 @@
...
+
+有两种计算二值分割任务的方法:
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:106]:
```suggestion
```

File: docs/zh_cn/faq.md [hunk 6]
```diff
@@ -66,3 +66,104 @@
...
+
+- 当 `out_channels=2` 时, 在训练时以 Cross Entropy Loss 作为损失函数, 在推理时使用 `F.softmax()` 归一化 logits 值, 然后通过 `argmax()` 得到每个像素的预测结果.
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:108]:
```suggestion
- 设置 `out_channels=2`, 在训练时以 Cross Entropy Loss 作为损失函数, 在推理时使用 `F.softmax()` 归一化 logits 值, 然后通过 `argmax()` 得到每个像素的预测结果.
```

File: docs/zh_cn/faq.md [hunk 7]
```diff
@@ -66,3 +66,104 @@
...
+
+- 当 `out_channels=1` 时, 我们在 [#2016](https://github.com/open-mmlab/mmsegmentation/pull/2016) 里提供了阈值参数 `threshold (默认为 0.3)`, 在训练时以 Binary Cross Entropy Loss 作为损失函数, 在推理时使用 `F.sigmoid()` 和 `threshold` 得到预测结果.
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:110]:
```suggestion
- 设置 `out_channels=1`, 在训练时以 Binary Cross Entropy Loss 作为损失函数, 在推理时使用 `F.sigmoid()` 和 `threshold` 得到预测结果, `threshold`的默认值为 0.3.
```

File: docs/zh_cn/faq.md [hunk 8]
```diff
@@ -66,3 +66,104 @@
...
+    seg_logit = F.softmax(seg_logit, dim=1)
+
+...
+
+if self.out_channels == 1:
+    seg_pred = (seg_logit >
+                self.decode_head.threshold).to(seg_logit).squeeze(1)
+else:
+    seg_pred = seg_logit.argmax(dim=1)
+```
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:126]: remove these lines

File: docs/zh_cn/faq.md [hunk 9]
```diff
@@ -66,3 +66,104 @@
...
+    loss_decode=dict(
+        type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0)),
+auxiliary_head=dict(
+    type='FCNHead',
+    in_channels=128,
+    in_index=3,
+    num_classes=2,
+    out_channels=1,
+    loss_decode=dict(
+        type='CrossEntropyLoss', use_sigmoid=True, loss_weight=0.4)),
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:154]: remove these lines

File: docs/zh_cn/faq.md [hunk 10]
```diff
@@ -66,3 +66,104 @@
...
+...
+
+if self.out_channels == 1:
+    seg_pred = (seg_logit >
+                self.decode_head.threshold).to(seg_logit).squeeze(1)
+else:
+    seg_pred = seg_logit.argmax(dim=1)
+```
+
+更多关于计算语义分割预测的细节可以参考 [encoder_decoder.py](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py):
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:128]:
```suggestion
更多关于实现细节可以参考 [encoder_decoder.py](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py):
```

File: docs/zh_cn/faq.md [hunk 11]
```diff
@@ -66,3 +66,104 @@
...
+
+对于实现上述两种计算二值分割的方法, 需要在 `decode_head` 和 `auxiliary_head` 的配置里修改:
+
+- (1) `num_classes=2`, `out_channels=2` 并在 `CrossEntropyLoss` 里面设置 `use_sigmoid=False`
+
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:133]: please add a cfg like 
```suggestion
```python
decode_head = dict( ..., out_channels=1, loss_decode=dict(
        type='CrossEntropyLoss', use_sigmoid=False, ...))
```
```

File: docs/zh_cn/faq.md [hunk 12]
```diff
@@ -66,3 +66,104 @@
...
+- (2) `num_classes=2`, `out_channels=1` 并在 `CrossEntropyLoss` 里面设置 `use_sigmoid=True`.
+
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:135]: also add a cfg example

File: docs/zh_cn/faq.md [hunk 13]
```diff
@@ -66,3 +66,104 @@
...
+如果采用解决方案 (2), 下面是对样例 [pspnet_unet_s5-d16.py](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/_base_/models/pspnet_unet_s5-d16.py) 做出的对应修改:
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:136]: I suggest add two examples for both of two method

File: docs/zh_cn/faq.md [hunk 14]
```diff
@@ -66,3 +66,104 @@
...
+
+```python
+if self.reduce_zero_label:
+    # avoid using underflow conversion
+    gt_semantic_seg[gt_semantic_seg == 0] = 255
+    gt_semantic_seg = gt_semantic_seg - 1
+    gt_semantic_seg[gt_semantic_seg == 254] = 255
+```
+
+`reduce_zero_label` 常常被用来处理 label 0 是背景的数据集, 如果 `reduce_zero_label=True`, label 0 对应的像素将不会参与损失函数的计算. 需要说明的是在二值分割任务中没有必要设置 `reduce_zero_label=True`, 请采用上面我们提到的解决方案.
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:169]:
```suggestion
数据集中 `reduce_zero_label`  参数的功能是为了忽略数据集 label 0。具体做法是，将 label 0 改为 255，其余 label 相应编号减1，同时 decode head 里 将 255 设为 ignore index，即不参与 loss 计算。 

以下是 `reduce_zero_label` 具体实现逻辑
```python
if self.reduce_zero_label:
    # avoid using underflow conversion
    gt_semantic_seg[gt_semantic_seg == 0] = 255
    gt_semantic_seg = gt_semantic_seg - 1
    gt_semantic_seg[gt_semantic_seg == 254] = 255
```

需要注意的是，在使用 `reduce_zero_label` 请确认数据集原始类别个数，如果只有两类，需要关闭 `reduce_zero_label` 即设置 `reduce_zero_label=False`.
```

File: docs/zh_cn/faq.md [hunk 15]
```diff
@@ -66,3 +66,74 @@
...
+    num_classes=2,
+    out_channels=1,
+    loss_decode=dict(
+        type='CrossEntropyLoss', use_sigmoid=True, loss_weight=0.4)),
+```
+
+## `reduce_zero_label` 的作用
+
+数据集中 `reduce_zero_label`参数类型为布尔类型, 默认为 False, 它的功能是为了忽略数据集 label 0. 具体做法是将 label 0 改为 255, 其余 label 相应编号减1, 同时 decode head 里将 255 设为 ignore index, 即不参与 loss 计算.
+以下是 `reduce_zero_label` 具体实现逻辑:
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:129]:
```suggestion
数据集中 `reduce_zero_label` 参数类型为布尔类型, 默认为 False, 它的功能是为了忽略数据集 label 0. 具体做法是将 label 0 改为 255, 其余 label 相应编号减 1, 同时 decode head 里将 255 设为 ignore index, 即不参与 loss 计算.
以下是 `reduce_zero_label` 具体实现逻辑:
```

File: docs/zh_cn/faq.md [hunk 16]
```diff
@@ -66,3 +66,72 @@
...
+
+```python
+if self.reduce_zero_label:
+    # avoid using underflow conversion
+    gt_semantic_seg[gt_semantic_seg == 0] = 255
+    gt_semantic_seg = gt_semantic_seg - 1
+    gt_semantic_seg[gt_semantic_seg == 254] = 255
+```
+
+需要注意的是, 使用 `reduce_zero_label` 请确认数据集原始类别个数, 如果只有两类, 需要关闭 `reduce_zero_label` 即设置 `reduce_zero_label=False`.
```
Comment (MeowZheng) [inline docs/zh_cn/faq.md:137]:
```suggestion
**注意:** 使用 `reduce_zero_label` 请确认数据集原始类别个数, 如果只有两类, 需要关闭 `reduce_zero_label` 即设置 `reduce_zero_label=False`.
```
